src/Scrutor/ServiceCollectionExtensions.Decoration.cs

```diff
@@ -21,7 +21,35 @@ public static partial class ServiceCollectionExtensions
         {
             Preconditions.NotNull(services, nameof(services));
 
-            return services.DecorateDescriptors(typeof(TService), x => x.Decorate(typeof(TDecorator)));
+            if (typeof(TDecorator).IsInterface)
+            {
+                return DecorateUsingInterface<TService, TDecorator>(services);
+            }
+            else
+            {
+                return services.DecorateDescriptors(typeof(TService), x => x.Decorate(typeof(TDecorator)));
+            }
+        }
+
+        private static IServiceCollection DecorateUsingInterface<TService, TDecorator>(IServiceCollection services) where TDecorator : TService
+        {
+            if (typeof(TDecorator).IsGenericType)
+            {
+                var decoratorDescriptor = services.Where(service => HasSameTypeDefinition(service.ServiceType, typeof(TDecorator))).FirstOrDefault();
+                if (decoratorDescriptor == null)
+                    throw new MissingTypeRegistrationException(typeof(TDecorator).IsGenericType ? typeof(TDecorator).GetGenericTypeDefinition() : typeof(TDecorator));
```

Comment:
Isn't `typeof(TDecorator).IsGenericType` always `true` here?

Reply:
Could you add some braces while you're at it? I like to always use braces to avoid GOTO FAIL-type bugs 😉  https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/

Reply:
> `IsGenericType`

not necessary, interfaces could be either generic or not

Reply:
> Could you add some braces while you're at it? I like to always use braces to avoid GOTO FAIL-type bugs 😉 https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/

totally, agree :)
thanks for pointing out